internal/controller/reinstall/reinstall.go

```diff
@@ -0,0 +1,378 @@
+func (r *ReinstallHandler) ensureValidReinstallRequest(
+	ctx context.Context,
+	log *zap.Logger,
+	clusterInstance *v1alpha1.ClusterInstance,
+) (result reconcile.Result, err error) {
+
+	result = reconcile.Result{}
+	err = nil
+
+	patch := client.MergeFrom(clusterInstance.DeepCopy())
+
+	var reinstallRequestCondition *metav1.Condition
+
+	if !r.ConfigStore.GetAllowReinstalls() {
+		log.Warn("SiteConfig Operator is not configured for cluster reinstalls")
+
+		err = fmt.Errorf("siteConfig operator is not configured for cluster reinstalls")
+		reinstallRequestCondition = reinstallRequestValidatedConditionStatus(metav1.ConditionFalse, v1alpha1.Failed,
+			"Cluster Reinstallation is not enabled")
+	} else {
+		reinstallRequestCondition = reinstallRequestValidatedConditionStatus(metav1.ConditionTrue, v1alpha1.Completed,
+			"Valid reinstall request")
+	}
+
+	updateRequired := false
+	if clusterInstance.Status.Reinstall.RequestStartTime.IsZero() {
+		clusterInstance.Status.Reinstall.RequestStartTime = metav1.Now()
+		updateRequired = true
+	}
+
+	if changed := setReinstallStatusCondition(clusterInstance, *reinstallRequestCondition); changed {
+		updateRequired = true
+	}
+
+	if updateRequired {
+		result = reconcile.Result{Requeue: true}
+		updateErr := patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch)
+		if updateErr != nil {
+			errMsg := "failed to update reinstall request validation status"
+			log.Error(errMsg, zap.Error(updateErr))
+			if err == nil {
+				err = updateErr
+			} else {
+				err = fmt.Errorf("%s (%w), err: %w", errMsg, updateErr, err)
+			}
+		}
+	}
+
+	return
+}
+
+// ensureRenderedManifestsAreDeleted deletes rendered manifests except for ManagedCluster
+func (r *ReinstallHandler) ensureRenderedManifestsAreDeleted(ctx context.Context, log *zap.Logger,
+	clusterInstance *v1alpha1.ClusterInstance) (result reconcile.Result, err error) {
+
+	result = reconcile.Result{}
+	err = nil
+
+	patch := client.MergeFrom(clusterInstance.DeepCopy())
+
+	renderedManifestsDeletedCondition := findReinstallStatusCondition(clusterInstance,
+		v1alpha1.ReinstallRenderedManifestsDeleted)
+
+	defer func() {
+		if renderedManifestsDeletedCondition != nil {
+			if changed := setReinstallStatusCondition(clusterInstance, *renderedManifestsDeletedCondition); changed {
+				result.Requeue = true
+				updateErr := patchClusterInstanceReinstallStatus(ctx, r.Client, log, clusterInstance, patch)
+				if updateErr != nil {
+					if err == nil {
+						err = updateErr
+					}
+				}
+			}
+		}
+	}()
+
+	excludeFromDeletion := make([]ci.RenderedObject, 1)
+	mc, err := getManagedCluster(clusterInstance)
+	if err != nil {
+		log.Error("Failed to retrieve ManagedCluster", zap.Error(err))
+		renderedManifestsDeletedCondition = reinstallRenderedManifestsDeletedConditionStatus(
+			metav1.ConditionFalse, v1alpha1.Failed, err.Error())
+		return
+	}
+
+	if mc != nil {
+		excludeFromDeletion[0] = *mc
```

Comment:
This feels like a strange way to handle this, but I suppose it's not wrong.
I would have went with:
```
var excludeFromDeletion []ci.RenderedObject
...
if mc != nil {
    excludeFromDeletion = []ci.RenderedOjbect{*mc}
}
```

Up to you if you want to switch though,